plasmapy/particles/tests/test_exceptions.py

```diff
@@ -0,0 +1,581 @@
+    (
+        "not normalized",
+        IonizationStateCollection,
+        [],
+        {"inputs": {"He": [0.4, 0.5, 0.0]}, "tol": 1e-9},
+        ParticleError,
+    ),
+    (
+        "negative ionfrac",
+        IonizationStateCollection,
+        [],
+        {"inputs": {"H": [-0.1, 1.1]}},
+        ParticleError,
+    ),
+    (
+        "ion",
+        IonizationStateCollection,
+        [],
+        {"inputs": {"H": [0.1, 0.9], "He+": [0.0, 0.9, 0.1]}},
+        ParticleError,
+    ),
+    (
+        "repeat elements",
+        IonizationStateCollection,
+        [],
+        {"inputs": {"H": [0.1, 0.9], "hydrogen": [0.2, 0.8]}},
+        ParticleError,
+    ),
+    (
+        "isotope of element",
+        IonizationStateCollection,
+        [],
+        {"inputs": {"H": [0.1, 0.9], "D": [0.2, 0.8]}},
+        ParticleError,
+    ),
+    (
+        "negative abundance",
+        IonizationStateCollection,
+        [],
+        {
+            "inputs": {"H": [0.1, 0.9], "He": [0.4, 0.5, 0.1]},
+            "abundances": {"H": 1, "He": -0.1},
+        },
+        ParticleError,
+    ),
+    (
+        "imaginary abundance",
+        IonizationStateCollection,
+        [],
+        {
+            "inputs": {"H": [0.1, 0.9], "He": [0.4, 0.5, 0.1]},
+            "abundances": {"H": 1, "He": 0.1j},
+        },
+        ParticleError,
+    ),
+    (
+        "wrong density units",
+        IonizationStateCollection,
+        [],
+        {
+            "inputs": {"H": [10, 90] * u.m ** -3, "He": [0.1, 0.9, 0] * u.m ** -2},
+            "abundances": {"H": 1, "He": 0.1},
+        },
+        ParticleError,
+    ),
+    (
+        "abundance redundance",
+        IonizationStateCollection,
+        [],
+        {
+            "inputs": {"H": [10, 90] * u.m ** -3, "He": [0.1, 0.9, 0] * u.m ** -3},
+            "abundances": {"H": 1, "He": 0.1},
+        },
+        ParticleError,
+    ),
+    (
+        "abundance contradiction",
+        IonizationStateCollection,
+        [],
+        {
+            "inputs": {"H": [10, 90] * u.m ** -3, "He": [0.1, 0.9, 0] * u.m ** -3},
+            "abundances": {"H": 1, "He": 0.11},
+        },
+        ParticleError,
+    ),
+    (
+        "kappa too small",
+        IonizationStateCollection,
+        [],
+        {"inputs": ["H"], "kappa": 1.499999},
+        ParticleError,
+    ),
+    (
+        "negative n",
+        IonizationStateCollection,
+        [],
+        {"inputs": ["H"], "n0": -1 * u.cm ** -3},
+        ParticleError,
+    ),
+    (
+        "negative T_e",
+        IonizationStateCollection,
+        [],
+        {"inputs": ["H-1"], "T_e": -1 * u.K},
+        ParticleError,
+    ),
+]
+
+tests_for_exceptions = (
+    tests_for_exceptions_IonizationState
+    + tests_for_exceptions_IonizationStateCollection
+)
+
+
+@pytest.mark.parametrize(
+    ["test_name", "tested_object", "args", "kwargs", "expected_exception"],
+    tests_for_exceptions,
+)
+def test_named_tests_for_exceptions(
```

Comment:
I see your point, but if you're changing a function's behavior by a lot, you're likely going to get failing tests in that other file :) I just grep to figure out where the tests are located.

And sure, I would prefer to keep them in one file but like you said - rewriting this kind of test in multiple places doesn't really make much sense. So I think it's fine to make an exception for the exceptions.

Reply:
One thing I skipped mentioning the first time around, but I keep thinking about it... while the main suite of tests for, say, Particles, lives, of course, in `particles/tests`, Particles are also indirectly tested from elsewhere in the package (everywhere that uses `@particle_input`, for example - so, most of the formulary)! So the assumption that all tests for code A live close to  code A is only correct until you have code B that depends on A - then B indirectly tests A. And, unless we have more leaves on the ~~vine~~ dependency graph than interconnected nodes, you'll have some B for most functionality pretty often.

Reply:
Totally agree all the tests for a given class/function should all be contained in the same file!  I thought this was always the case for us.  Do we have tests that are split and, if so, which ones are they?

I also find it distracting if there all multiple tests functions for a given function/class.  It makes reading the test file very difficult.  This is a big reason why most of the test I write are combined into one test class.  I find the implicit grouping of a class makes navigating the test code much easier, especially with a good IDE.

Now, I generally don't like to test exceptions, warnings, and valid calls all in the same test function, but I do apply this on a case-by-cases basis.  If tests are short and can be easily be done in one test function, then I'll combine them all.  Otherwise, I like separating them because then I don't have to mentally switch between "what I'm testing."  Since I don't have to do the mental switch, interpreting the test failures becomes easier.  Additionally, for me, writing exception tests first, then warning tests, and then valid call tests allows me to mentally walk through my code to get good coverage and really debug how it's behaving.  Plus, I think it fits well into the idea of modularization you code.  It's all a bit of a art form and a personal preference.